dftimewolf/lib/processors/gcp_cloud_resource_tree.py

```diff
@@ -0,0 +1,1199 @@
+# -*- coding: utf-8 -*-
+"""Creates a GCP cloud resource tree."""
+
+import json
+import tempfile
+from datetime import datetime
+from datetime import timedelta
+from datetime import timezone
+from typing import Dict, List, Optional, Any, Set
+from enum import Enum, auto
+
+from dateutil import parser
+from libcloudforensics.providers.gcp.internal.common import CreateService
+from libcloudforensics.providers.gcp.internal.log import GoogleCloudLog
+
+from dftimewolf.lib import module
+from dftimewolf.lib.containers import containers
+from dftimewolf.lib.modules import manager as modules_manager
+from dftimewolf.lib.state import DFTimewolfState
+
+
+class OperatingMode(Enum):
+  """Enum represent operational mode (Online or Offline)."""
+
+  ONLINE = auto()
+  OFFLINE = auto()
+
+
+class Resource:
+  """A Class that represent a resource (Instance, Disk, Image...etc)."""
+
+  has_dynamic_attributes = True  # silences all attribute-errors for Resource
+
+  def __init__(self) -> None:
+    """Initialize the Resource object."""
+    self.id = ''
+    self.name = ''
+
+    self.type = ''
+    self.state = ''
+    self.project_id = ''
+    self.zone = ''
+
+    self.created_by = ''
+    self.creator_ip_address = ''
+    self.creator_useragent = ''
+    self.deleted_by = ''
+    self.deleter_ip_address = ''
+    self.deleter_useragent = ''
+    self.parent: Optional[Resource] = None
+    self.children: Set[Resource] = set()
+    self.disks: List[Resource] = []
+    self.deleted = False
+    self._resource_name = ''
+    self._creation_timestamp: Optional[datetime] = None
+    self._deletion_timestamp: Optional[datetime] = None
+
+  def __hash__(self) -> int:
+    """For object comparison."""
+    return hash(self.id + self.resource_name)
+
+  @property
+  def resource_name(self) -> str:
+    """Property resource_name."""
+    return self._resource_name
+
+  @resource_name.setter
+  def resource_name(self, value: str) -> None:
```

Comment:
it took me a while to understand where does value come from and how it looks like, my understanding is that it should be equal to the resource_name log field (either jsonPayload.resource.name or protoPayload.resourceName), please add docs mentioning this and a link to the docs https://cloud.google.com/compute/docs/logging/migrating-from-activity-logs-to-audit-logs#fields


)

Reply:
resource_name is a property of Resource. "value" is the value you set it to for example resource.resource_name = "anything".

I added a comment with examples of how value is expected.